### Hyper-Parameter Tuning with CV

This notebook will cover exercise answer.

* Exercise 9.1
* Exercise 9.2
* Exercise 9.3
* Exercise 9.4
* Exercise 9.5
* Exercise 9.6

As we go along, there will be some explanations.

Most of the functions below can be found under Tool/metrics

Contact: boyboi86@gmail.com

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import research as rs

%matplotlib inline

Num of CPU core:  4
Machine info:  Windows-10-10.0.18362-SP0
Python 3.7.4 (default, Aug  9 2019, 18:34:13) [MSC v.1915 64 bit (AMD64)]
Numpy 1.17.3
Pandas 1.0.3


C:\Users\Wei_X\Anaconda3\lib\site-packages\statsmodels\tools\_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


<Figure size 1500x800 with 1 Axes>

In [2]:
X, y = rs.make_data(n_features=10, n_informative=5, n_redundant= 0, n_samples= 2000)

#Take note of depreciation warning Pandas 1.0.3

C:\Users\Wei_X\Desktop\Python\research\Util\indicator.py:54: FutureWarning: The pandas.datetime class is deprecated and will be removed from pandas in a future version. Import from datetime module instead.
  df0 = pd.date_range(periods=n_samples, freq=pd.tseries.offsets.BDay(), end=pd.datetime.today())


In [3]:
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
from sklearn.svm import SVC

n_splits = 10

param_grid = {'kernel':['rbf'], #redundant; default is already 'rbf'
              'C':[1e-2, 1e-1, 1, 10, 100],
              'gamma': [1e-2, 1e-1, 1, 10, 100]}

clf_svm = SVC(kernel='rbf',
              probability=True)

clf = GridSearchCV(estimator = clf_svm, 
                   param_grid = param_grid,
                   scoring='neg_log_loss', 
                   n_jobs=None, 
                   #iid='deprecated', take note of depreciation! version 0.22 warning
                   refit=True,
                   cv= n_splits, # 10 fold CV, but! stratified KFold
                   verbose=10)

clf.fit(X,
        y['bin'],
       sample_weight = None) #sample_weight here!!



[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


Fitting 10 folds for each of 25 candidates, totalling 250 fits
[CV] C=0.01, gamma=0.01, kernel=rbf ..................................
[CV] ..... C=0.01, gamma=0.01, kernel=rbf, score=-0.287, total=   0.6s
[CV] C=0.01, gamma=0.01, kernel=rbf ..................................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.5s remaining:    0.0s


[CV] ..... C=0.01, gamma=0.01, kernel=rbf, score=-0.260, total=   0.6s
[CV] C=0.01, gamma=0.01, kernel=rbf ..................................


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    1.0s remaining:    0.0s


[CV] ..... C=0.01, gamma=0.01, kernel=rbf, score=-0.198, total=   0.5s
[CV] C=0.01, gamma=0.01, kernel=rbf ..................................


[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:    1.6s remaining:    0.0s


[CV] ..... C=0.01, gamma=0.01, kernel=rbf, score=-0.186, total=   0.5s
[CV] C=0.01, gamma=0.01, kernel=rbf ..................................


[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed:    2.1s remaining:    0.0s


[CV] ..... C=0.01, gamma=0.01, kernel=rbf, score=-0.271, total=   0.5s
[CV] C=0.01, gamma=0.01, kernel=rbf ..................................


[Parallel(n_jobs=1)]: Done   5 out of   5 | elapsed:    2.6s remaining:    0.0s


[CV] ..... C=0.01, gamma=0.01, kernel=rbf, score=-0.618, total=   0.5s
[CV] C=0.01, gamma=0.01, kernel=rbf ..................................


[Parallel(n_jobs=1)]: Done   6 out of   6 | elapsed:    3.2s remaining:    0.0s


[CV] ..... C=0.01, gamma=0.01, kernel=rbf, score=-0.515, total=   0.5s
[CV] C=0.01, gamma=0.01, kernel=rbf ..................................


[Parallel(n_jobs=1)]: Done   7 out of   7 | elapsed:    3.7s remaining:    0.0s


[CV] ..... C=0.01, gamma=0.01, kernel=rbf, score=-0.605, total=   0.5s
[CV] C=0.01, gamma=0.01, kernel=rbf ..................................


[Parallel(n_jobs=1)]: Done   8 out of   8 | elapsed:    4.2s remaining:    0.0s


[CV] ..... C=0.01, gamma=0.01, kernel=rbf, score=-0.507, total=   0.5s
[CV] C=0.01, gamma=0.01, kernel=rbf ..................................


[Parallel(n_jobs=1)]: Done   9 out of   9 | elapsed:    4.8s remaining:    0.0s


[CV] ..... C=0.01, gamma=0.01, kernel=rbf, score=-0.623, total=   0.5s
[CV] C=0.01, gamma=0.1, kernel=rbf ...................................
[CV] ...... C=0.01, gamma=0.1, kernel=rbf, score=-0.276, total=   0.5s
[CV] C=0.01, gamma=0.1, kernel=rbf ...................................
[CV] ...... C=0.01, gamma=0.1, kernel=rbf, score=-0.231, total=   0.5s
[CV] C=0.01, gamma=0.1, kernel=rbf ...................................
[CV] ...... C=0.01, gamma=0.1, kernel=rbf, score=-0.202, total=   0.5s
[CV] C=0.01, gamma=0.1, kernel=rbf ...................................
[CV] ...... C=0.01, gamma=0.1, kernel=rbf, score=-0.205, total=   0.5s
[CV] C=0.01, gamma=0.1, kernel=rbf ...................................
[CV] ...... C=0.01, gamma=0.1, kernel=rbf, score=-0.223, total=   0.5s
[CV] C=0.01, gamma=0.1, kernel=rbf ...................................
[CV] ...... C=0.01, gamma=0.1, kernel=rbf, score=-0.545, total=   0.5s
[CV] C=0.01, gamma=0.1, kernel=rbf ...................................
[CV] .

[CV] ....... C=0.1, gamma=0.1, kernel=rbf, score=-0.342, total=   0.3s
[CV] C=0.1, gamma=0.1, kernel=rbf ....................................
[CV] ....... C=0.1, gamma=0.1, kernel=rbf, score=-0.346, total=   0.3s
[CV] C=0.1, gamma=0.1, kernel=rbf ....................................
[CV] ....... C=0.1, gamma=0.1, kernel=rbf, score=-0.433, total=   0.3s
[CV] C=0.1, gamma=1, kernel=rbf ......................................
[CV] ......... C=0.1, gamma=1, kernel=rbf, score=-0.693, total=   0.5s
[CV] C=0.1, gamma=1, kernel=rbf ......................................
[CV] ......... C=0.1, gamma=1, kernel=rbf, score=-0.694, total=   0.5s
[CV] C=0.1, gamma=1, kernel=rbf ......................................
[CV] ......... C=0.1, gamma=1, kernel=rbf, score=-0.693, total=   0.5s
[CV] C=0.1, gamma=1, kernel=rbf ......................................
[CV] ......... C=0.1, gamma=1, kernel=rbf, score=-0.695, total=   0.5s
[CV] C=0.1, gamma=1, kernel=rbf ......................................
[CV] .

[CV] ........... C=1, gamma=1, kernel=rbf, score=-0.614, total=   0.6s
[CV] C=1, gamma=1, kernel=rbf ........................................
[CV] ........... C=1, gamma=1, kernel=rbf, score=-0.522, total=   0.6s
[CV] C=1, gamma=1, kernel=rbf ........................................
[CV] ........... C=1, gamma=1, kernel=rbf, score=-0.527, total=   0.7s
[CV] C=1, gamma=1, kernel=rbf ........................................
[CV] ........... C=1, gamma=1, kernel=rbf, score=-0.675, total=   0.7s
[CV] C=1, gamma=1, kernel=rbf ........................................
[CV] ........... C=1, gamma=1, kernel=rbf, score=-0.551, total=   0.7s
[CV] C=1, gamma=10, kernel=rbf .......................................
[CV] .......... C=1, gamma=10, kernel=rbf, score=-0.693, total=   0.8s
[CV] C=1, gamma=10, kernel=rbf .......................................
[CV] .......... C=1, gamma=10, kernel=rbf, score=-0.693, total=   0.7s
[CV] C=1, gamma=10, kernel=rbf .......................................
[CV] .

[CV] ......... C=10, gamma=10, kernel=rbf, score=-0.693, total=   0.8s
[CV] C=10, gamma=10, kernel=rbf ......................................
[CV] ......... C=10, gamma=10, kernel=rbf, score=-0.693, total=   0.7s
[CV] C=10, gamma=10, kernel=rbf ......................................
[CV] ......... C=10, gamma=10, kernel=rbf, score=-0.693, total=   0.8s
[CV] C=10, gamma=10, kernel=rbf ......................................
[CV] ......... C=10, gamma=10, kernel=rbf, score=-0.693, total=   0.7s
[CV] C=10, gamma=10, kernel=rbf ......................................
[CV] ......... C=10, gamma=10, kernel=rbf, score=-0.693, total=   0.8s
[CV] C=10, gamma=10, kernel=rbf ......................................
[CV] ......... C=10, gamma=10, kernel=rbf, score=-0.693, total=   0.9s
[CV] C=10, gamma=10, kernel=rbf ......................................
[CV] ......... C=10, gamma=10, kernel=rbf, score=-0.693, total=   0.8s
[CV] C=10, gamma=100, kernel=rbf .....................................
[CV] .

[CV] ....... C=100, gamma=100, kernel=rbf, score=-0.693, total=   1.4s
[CV] C=100, gamma=100, kernel=rbf ....................................
[CV] ....... C=100, gamma=100, kernel=rbf, score=-0.693, total=   1.3s
[CV] C=100, gamma=100, kernel=rbf ....................................
[CV] ....... C=100, gamma=100, kernel=rbf, score=-0.693, total=   1.3s
[CV] C=100, gamma=100, kernel=rbf ....................................
[CV] ....... C=100, gamma=100, kernel=rbf, score=-0.693, total=   1.6s
[CV] C=100, gamma=100, kernel=rbf ....................................
[CV] ....... C=100, gamma=100, kernel=rbf, score=-0.693, total=   1.4s
[CV] C=100, gamma=100, kernel=rbf ....................................
[CV] ....... C=100, gamma=100, kernel=rbf, score=-0.693, total=   1.3s
[CV] C=100, gamma=100, kernel=rbf ....................................
[CV] ....... C=100, gamma=100, kernel=rbf, score=-0.693, total=   1.3s
[CV] C=100, gamma=100, kernel=rbf ....................................
[CV] .

[Parallel(n_jobs=1)]: Done 250 out of 250 | elapsed:  2.8min finished


GridSearchCV(cv=10, estimator=SVC(probability=True),
             param_grid={'C': [0.01, 0.1, 1, 10, 100],
                         'gamma': [0.01, 0.1, 1, 10, 100], 'kernel': ['rbf']},
             scoring='neg_log_loss', verbose=10)

**Note 1**

I allowed verbose = 10, so that we can see the output messages. Based on my due diligence and educated guess from these code snippets in the book, I believe Dr Marco was running sklearn version 0.14 - 0.16.

The only thing I found online for sklearn node_count was for decision tree, hence the answer provided is experimental (I think that was what Dr Marco wanted?).

[sklearn SVC](https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html?highlight=svc#sklearn.svm.SVC)

[sklearn GridSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html?highlight=gridsearchcv#sklearn.model_selection.GridSearchCV)

[sklearn hyper-parameter tuning](https://scikit-learn.org/stable/modules/grid_search.html#gridsearch-scoring)

The upside of using sklearn 0.23.1, most of the function found in the book can be found in their current library.

* uniform log
* sample_weight can be passed to fit method

**Note 2**

As for how I derived combination of param_grids as "nodes" in the below answers. Kindly refer to [ML Mastery](https://machinelearningmastery.com/how-to-configure-the-number-of-layers-and-nodes-in-a-neural-network/#:~:text=A%20node%2C%20also%20called%20a,layers%20to%20comprise%20a%20network.)

The below is a diagram which I found from google.

![logo](https://pathmind.com/images/wiki/perceptron_node.png)

Although this is under Deep-ML, but given the time of writing we will have to perform an educated guess to match the author's intention as well as terms mentioned in the book (Match functionality to form connections).

Essentially, nodes are perceptors (Deep-Learning) which are **inputs**. In our case, Grid Search (GS) inputs are param_grids.

Hence, I believe this assumption does make sense in this GS exercise's context.

**Note 3**

At the same time, The results I placed below is mostly average/ mean. The reality is after splitting. Each node/ candidate were ran once on each sample.

In this example, I only generated 2,000 samples.

In [4]:
def print_result(clf, n_splits):
    num_nodes = len(clf.cv_results_['rank_test_score'])
    total_time = clf.refit_time_ * num_nodes * n_splits

    best_scores, best_scores_idx, mean_score_time, mean_fit_time = [], 0, .0, .0

    for i in np.arange(n_splits):
        best_scores.append(clf.cv_results_["split"+ str(i) +"_test_score"][clf.best_index_])
        idx = np.where(max(best_scores) == best_scores)[0][0] #always + 1 because index starts from 0 as default

    best_scores_idx = (clf.best_index_ + 1) + len(clf.cv_results_['mean_score_time']) * idx
    print("Best params for estimator: {0}\nBest CV Score: {1:.6f}\n".format(clf.best_estimator_, max(best_scores)))
    print("A total of {0} was performed before optimal CV score found! (Under split{1}_test_score / {2}th split)\n".format(best_scores_idx, idx, idx + 1))

    print("Estimated time taken for entire process: {0:.6f} seconds\nTotal number of candidates/ nodes: {1}\n".format(total_time, 
                                                                                                           num_nodes))
    i = 0
    while i < (clf.best_index_ + 1):
        mean_score_time += clf.cv_results_['mean_score_time'][i]
        mean_fit_time += clf.cv_results_['mean_fit_time'][i]
        i+=1
    print("=" * 55)
    print("Estimated total mean time required for optimal solution:\n\nScore Time: {0:.6f}s\nFit Time: {1:.6f}s".format(mean_score_time, mean_fit_time))

print_result(clf = clf, n_splits = n_splits)

Best params for estimator: SVC(C=1, gamma=0.1, probability=True)
Best CV Score: -0.098618

A total of 87 was performed before optimal CV score found! (Under split3_test_score / 4th split)

Estimated time taken for entire process: 74.261129 seconds
Total number of candidates/ nodes: 25

Estimated total mean time required for optimal solution:

Score Time: 0.160910s
Fit Time: 6.813837s


In [5]:
from sklearn.utils.fixes import loguniform

param_distributions = {'C':loguniform(a = 1e-2, b= 1e2),
                      'gamma':loguniform(a = 1e-2, b= 1e2)}

Rclf = RandomizedSearchCV(estimator = clf_svm, 
                          param_distributions = param_distributions,
                          n_iter = 25, # newly added as well
                          scoring='neg_log_loss', 
                          n_jobs=None, 
                          #iid='deprecated', take note of depreciation! version 0.22 warning
                          refit=True,
                          cv= n_splits, # 10 fold CV, but! stratified KFold
                          verbose=10)

Rclf.fit(X,
         y['bin'],
         sample_weight = None) #sample_weight here!!

Fitting 10 folds for each of 25 candidates, totalling 250 fits
[CV] C=0.7833260760128987, gamma=0.2994822977403476 ..................


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV]  C=0.7833260760128987, gamma=0.2994822977403476, score=-0.271, total=   0.5s
[CV] C=0.7833260760128987, gamma=0.2994822977403476 ..................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.4s remaining:    0.0s


[CV]  C=0.7833260760128987, gamma=0.2994822977403476, score=-0.158, total=   0.5s
[CV] C=0.7833260760128987, gamma=0.2994822977403476 ..................


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    0.9s remaining:    0.0s


[CV]  C=0.7833260760128987, gamma=0.2994822977403476, score=-0.132, total=   0.5s
[CV] C=0.7833260760128987, gamma=0.2994822977403476 ..................


[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:    1.4s remaining:    0.0s


[CV]  C=0.7833260760128987, gamma=0.2994822977403476, score=-0.114, total=   0.5s
[CV] C=0.7833260760128987, gamma=0.2994822977403476 ..................


[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed:    1.9s remaining:    0.0s


[CV]  C=0.7833260760128987, gamma=0.2994822977403476, score=-0.163, total=   0.5s
[CV] C=0.7833260760128987, gamma=0.2994822977403476 ..................


[Parallel(n_jobs=1)]: Done   5 out of   5 | elapsed:    2.4s remaining:    0.0s


[CV]  C=0.7833260760128987, gamma=0.2994822977403476, score=-0.390, total=   0.5s
[CV] C=0.7833260760128987, gamma=0.2994822977403476 ..................


[Parallel(n_jobs=1)]: Done   6 out of   6 | elapsed:    2.9s remaining:    0.0s


[CV]  C=0.7833260760128987, gamma=0.2994822977403476, score=-0.378, total=   0.5s
[CV] C=0.7833260760128987, gamma=0.2994822977403476 ..................


[Parallel(n_jobs=1)]: Done   7 out of   7 | elapsed:    3.4s remaining:    0.0s


[CV]  C=0.7833260760128987, gamma=0.2994822977403476, score=-0.289, total=   0.5s
[CV] C=0.7833260760128987, gamma=0.2994822977403476 ..................


[Parallel(n_jobs=1)]: Done   8 out of   8 | elapsed:    3.9s remaining:    0.0s


[CV]  C=0.7833260760128987, gamma=0.2994822977403476, score=-0.377, total=   0.5s
[CV] C=0.7833260760128987, gamma=0.2994822977403476 ..................


[Parallel(n_jobs=1)]: Done   9 out of   9 | elapsed:    4.4s remaining:    0.0s


[CV]  C=0.7833260760128987, gamma=0.2994822977403476, score=-0.373, total=   0.5s
[CV] C=0.49037818303803193, gamma=8.536817927058488 ..................
[CV]  C=0.49037818303803193, gamma=8.536817927058488, score=-0.693, total=   0.6s
[CV] C=0.49037818303803193, gamma=8.536817927058488 ..................
[CV]  C=0.49037818303803193, gamma=8.536817927058488, score=-0.693, total=   0.6s
[CV] C=0.49037818303803193, gamma=8.536817927058488 ..................
[CV]  C=0.49037818303803193, gamma=8.536817927058488, score=-0.693, total=   0.7s
[CV] C=0.49037818303803193, gamma=8.536817927058488 ..................
[CV]  C=0.49037818303803193, gamma=8.536817927058488, score=-0.693, total=   0.7s
[CV] C=0.49037818303803193, gamma=8.536817927058488 ..................
[CV]  C=0.49037818303803193, gamma=8.536817927058488, score=-0.693, total=   0.8s
[CV] C=0.49037818303803193, gamma=8.536817927058488 ..................
[CV]  C=0.49037818303803193, gamma=8.536817927058488, score=-0.694, total=   0.6s


[CV]  C=35.48667662529329, gamma=4.403630972109106, score=-0.694, total=   0.7s
[CV] C=35.48667662529329, gamma=4.403630972109106 ....................
[CV]  C=35.48667662529329, gamma=4.403630972109106, score=-0.693, total=   0.6s
[CV] C=35.48667662529329, gamma=4.403630972109106 ....................
[CV]  C=35.48667662529329, gamma=4.403630972109106, score=-0.693, total=   0.6s
[CV] C=35.48667662529329, gamma=4.403630972109106 ....................
[CV]  C=35.48667662529329, gamma=4.403630972109106, score=-0.693, total=   0.6s
[CV] C=35.48667662529329, gamma=4.403630972109106 ....................
[CV]  C=35.48667662529329, gamma=4.403630972109106, score=-0.693, total=   0.6s
[CV] C=35.48667662529329, gamma=4.403630972109106 ....................
[CV]  C=35.48667662529329, gamma=4.403630972109106, score=-0.693, total=   0.7s
[CV] C=35.48667662529329, gamma=4.403630972109106 ....................
[CV]  C=35.48667662529329, gamma=4.403630972109106, score=-0.693, total=   0.6s
[CV] C=5.97872

[CV]  C=0.029600079668876515, gamma=0.6614167167922481, score=-0.694, total=   0.5s
[CV] C=0.029600079668876515, gamma=0.6614167167922481 ................
[CV]  C=0.029600079668876515, gamma=0.6614167167922481, score=-0.693, total=   0.5s
[CV] C=0.029600079668876515, gamma=0.6614167167922481 ................
[CV]  C=0.029600079668876515, gamma=0.6614167167922481, score=-0.694, total=   0.5s
[CV] C=0.27015114208486657, gamma=2.2135429049733886 .................
[CV]  C=0.27015114208486657, gamma=2.2135429049733886, score=-0.694, total=   0.6s
[CV] C=0.27015114208486657, gamma=2.2135429049733886 .................
[CV]  C=0.27015114208486657, gamma=2.2135429049733886, score=-0.693, total=   0.6s
[CV] C=0.27015114208486657, gamma=2.2135429049733886 .................
[CV]  C=0.27015114208486657, gamma=2.2135429049733886, score=-0.693, total=   0.6s
[CV] C=0.27015114208486657, gamma=2.2135429049733886 .................
[CV]  C=0.27015114208486657, gamma=2.2135429049733886, score=-0.694, tota

[CV]  C=0.8984964421366174, gamma=19.901842696875132, score=-0.694, total=   0.9s
[CV] C=0.8984964421366174, gamma=19.901842696875132 ..................
[CV]  C=0.8984964421366174, gamma=19.901842696875132, score=-0.693, total=   0.9s
[CV] C=0.8984964421366174, gamma=19.901842696875132 ..................
[CV]  C=0.8984964421366174, gamma=19.901842696875132, score=-0.693, total=   0.9s
[CV] C=0.8984964421366174, gamma=19.901842696875132 ..................
[CV]  C=0.8984964421366174, gamma=19.901842696875132, score=-0.693, total=   0.9s
[CV] C=0.8984964421366174, gamma=19.901842696875132 ..................
[CV]  C=0.8984964421366174, gamma=19.901842696875132, score=-0.695, total=   0.9s
[CV] C=0.8984964421366174, gamma=19.901842696875132 ..................
[CV]  C=0.8984964421366174, gamma=19.901842696875132, score=-0.693, total=   0.9s
[CV] C=0.8984964421366174, gamma=19.901842696875132 ..................
[CV]  C=0.8984964421366174, gamma=19.901842696875132, score=-0.694, total=   1.0s


[CV]  C=87.1657109323045, gamma=10.197495561031324, score=-0.693, total=   0.7s
[CV] C=87.1657109323045, gamma=10.197495561031324 ....................
[CV]  C=87.1657109323045, gamma=10.197495561031324, score=-0.693, total=   0.7s
[CV] C=87.1657109323045, gamma=10.197495561031324 ....................
[CV]  C=87.1657109323045, gamma=10.197495561031324, score=-0.693, total=   0.7s
[CV] C=87.1657109323045, gamma=10.197495561031324 ....................
[CV]  C=87.1657109323045, gamma=10.197495561031324, score=-0.693, total=   0.7s
[CV] C=87.1657109323045, gamma=10.197495561031324 ....................
[CV]  C=87.1657109323045, gamma=10.197495561031324, score=-0.693, total=   0.7s
[CV] C=19.930431094669313, gamma=1.4785219604311253 ..................
[CV]  C=19.930431094669313, gamma=1.4785219604311253, score=-0.492, total=   0.6s
[CV] C=19.930431094669313, gamma=1.4785219604311253 ..................
[CV]  C=19.930431094669313, gamma=1.4785219604311253, score=-0.456, total=   0.6s
[CV] C=19.

[Parallel(n_jobs=1)]: Done 250 out of 250 | elapsed:  2.5min finished


RandomizedSearchCV(cv=10, estimator=SVC(probability=True), n_iter=25,
                   param_distributions={'C': <scipy.stats._distn_infrastructure.rv_frozen object at 0x000001DCA53709C8>,
                                        'gamma': <scipy.stats._distn_infrastructure.rv_frozen object at 0x000001DCA439F148>},
                   scoring='neg_log_loss', verbose=10)

In [6]:
print_result(clf = Rclf, n_splits = n_splits)

Best params for estimator: SVC(C=1.308709482507712, gamma=0.06937936028678086, probability=True)
Best CV Score: -0.095234

A total of 85 was performed before optimal CV score found! (Under split3_test_score / 4th split)

Estimated time taken for entire process: 67.476273 seconds
Total number of candidates/ nodes: 25

Estimated total mean time required for optimal solution:

Score Time: 0.111877s
Fit Time: 5.103568s


### Comparison between GSCV vs RSCV:

**Grid Search CV**

Best params for estimator: SVC(C=1, gamma=0.1, probability=True)

Best CV Score: -0.095049

A total of 87 was performed before optimal CV score found! (Under split3_test_score / 4th split)

Estimated time taken for entire process: 77.301800 seconds
Total number of candidates/ nodes: 25

Estimated total mean time required for optimal solution:

Score Time: 0.156009s
Fit Time: 6.964366s

**Random Search CV**

Best params for estimator: SVC(C=5.146914740190327, gamma=0.03550270078366642, probability=True)

Best CV Score: -0.100946 (Lowest)

A total of 78 was performed before optimal CV score found! (Under split3_test_score / 4th split)

Estimated time taken for entire process: 62.176406 seconds
Total number of candidates/ nodes: 25

Estimated total mean time required for optimal solution:

Score Time: 0.021389s
Fit Time: 1.155873s

### Conclusion

Number of nodes are both 25 (RSCV: n_iter = 25), (GSCV = 5 * 5 inputs).

While RSCV does perform significantly much faster for both score and fit time as well as the number of fits required.

GSCV has a higher neg_log_loss score (Not optimal prediction score as compared to RSCV).

Hence RSCV does seem overall better with a lower score and faster results.

But as seen in previous chapters, randomness will usually cause an inflated score (more bias). Hence, GSCV may provide a more reliable outcome.

**Note**

For sklearn, their attributes after fit will always return the highest mean score (Worst score for neg_log_loss). Hence the above comparison is more like comparing between the worst log_loss reported, which method provides the better prediction score (lowest).

In [7]:
def in_sample_sharpe_ratio(clf):
    sharpe_ratio = []
    for i in np.arange(len(clf.cv_results_['mean_test_score'])):
        if clf.cv_results_['mean_test_score'][i] < 0:
            sharpe_ratio.append(-1 * clf.cv_results_['mean_test_score'][i]/ clf.cv_results_['std_test_score'][i])
        else:
            sharpe_ratio.append(clf.cv_results_['mean_test_score'][i]/ clf.cv_results_['std_test_score'][i])
    print("IS Best Score Sharpe Ratio: {0:.6f}".format(sharpe_ratio[clf.best_index_]))
    print("Best IS Sharpe ratio: {0:.6f}\nLowest IS Sharpe Ratio: {1:.6f}\nMean Sharpe Ratio: {2:.6f}".format(max(sharpe_ratio), min(sharpe_ratio), np.mean(sharpe_ratio)))

In [8]:
in_sample_sharpe_ratio(clf)

IS Best Score Sharpe Ratio: 2.487544
Best IS Sharpe ratio: inf
Lowest IS Sharpe Ratio: 2.194502
Mean Sharpe Ratio: inf


C:\Users\Wei_X\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: RuntimeWarning: divide by zero encountered in double_scalars
  """


In [9]:
clf1 = GridSearchCV(estimator = clf_svm, 
                   param_grid = param_grid,
                   scoring='accuracy', #change to accuracy
                   n_jobs=None, 
                   #iid='deprecated', take note of depreciation! version 0.22 warning
                   refit=True,
                   cv= n_splits, # 10 fold CV, but! stratified KFold
                   verbose=10)

clf1.fit(X,
        y['bin'],
       sample_weight = None) #sample_weight here!!

in_sample_sharpe_ratio(clf1)

Fitting 10 folds for each of 25 candidates, totalling 250 fits
[CV] C=0.01, gamma=0.01, kernel=rbf ..................................


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV] ...... C=0.01, gamma=0.01, kernel=rbf, score=0.890, total=   0.5s
[CV] C=0.01, gamma=0.01, kernel=rbf ..................................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.4s remaining:    0.0s


[CV] ...... C=0.01, gamma=0.01, kernel=rbf, score=0.880, total=   0.5s
[CV] C=0.01, gamma=0.01, kernel=rbf ..................................


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    0.9s remaining:    0.0s


[CV] ...... C=0.01, gamma=0.01, kernel=rbf, score=0.915, total=   0.5s
[CV] C=0.01, gamma=0.01, kernel=rbf ..................................


[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:    1.4s remaining:    0.0s


[CV] ...... C=0.01, gamma=0.01, kernel=rbf, score=0.945, total=   0.5s
[CV] C=0.01, gamma=0.01, kernel=rbf ..................................


[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed:    1.9s remaining:    0.0s


[CV] ...... C=0.01, gamma=0.01, kernel=rbf, score=0.900, total=   0.5s
[CV] C=0.01, gamma=0.01, kernel=rbf ..................................


[Parallel(n_jobs=1)]: Done   5 out of   5 | elapsed:    2.4s remaining:    0.0s


[CV] ...... C=0.01, gamma=0.01, kernel=rbf, score=0.700, total=   0.5s
[CV] C=0.01, gamma=0.01, kernel=rbf ..................................


[Parallel(n_jobs=1)]: Done   6 out of   6 | elapsed:    2.9s remaining:    0.0s


[CV] ...... C=0.01, gamma=0.01, kernel=rbf, score=0.775, total=   0.5s
[CV] C=0.01, gamma=0.01, kernel=rbf ..................................


[Parallel(n_jobs=1)]: Done   7 out of   7 | elapsed:    3.4s remaining:    0.0s


[CV] ...... C=0.01, gamma=0.01, kernel=rbf, score=0.685, total=   0.5s
[CV] C=0.01, gamma=0.01, kernel=rbf ..................................


[Parallel(n_jobs=1)]: Done   8 out of   8 | elapsed:    3.9s remaining:    0.0s


[CV] ...... C=0.01, gamma=0.01, kernel=rbf, score=0.735, total=   0.5s
[CV] C=0.01, gamma=0.01, kernel=rbf ..................................


[Parallel(n_jobs=1)]: Done   9 out of   9 | elapsed:    4.4s remaining:    0.0s


[CV] ...... C=0.01, gamma=0.01, kernel=rbf, score=0.715, total=   0.5s
[CV] C=0.01, gamma=0.1, kernel=rbf ...................................
[CV] ....... C=0.01, gamma=0.1, kernel=rbf, score=0.920, total=   0.5s
[CV] C=0.01, gamma=0.1, kernel=rbf ...................................
[CV] ....... C=0.01, gamma=0.1, kernel=rbf, score=0.920, total=   0.5s
[CV] C=0.01, gamma=0.1, kernel=rbf ...................................
[CV] ....... C=0.01, gamma=0.1, kernel=rbf, score=0.960, total=   0.5s
[CV] C=0.01, gamma=0.1, kernel=rbf ...................................
[CV] ....... C=0.01, gamma=0.1, kernel=rbf, score=0.970, total=   0.5s
[CV] C=0.01, gamma=0.1, kernel=rbf ...................................
[CV] ....... C=0.01, gamma=0.1, kernel=rbf, score=0.930, total=   0.5s
[CV] C=0.01, gamma=0.1, kernel=rbf ...................................
[CV] ....... C=0.01, gamma=0.1, kernel=rbf, score=0.795, total=   0.5s
[CV] C=0.01, gamma=0.1, kernel=rbf ...................................
[CV] .

[CV] ........ C=0.1, gamma=0.1, kernel=rbf, score=0.825, total=   0.3s
[CV] C=0.1, gamma=0.1, kernel=rbf ....................................
[CV] ........ C=0.1, gamma=0.1, kernel=rbf, score=0.860, total=   0.3s
[CV] C=0.1, gamma=0.1, kernel=rbf ....................................
[CV] ........ C=0.1, gamma=0.1, kernel=rbf, score=0.820, total=   0.3s
[CV] C=0.1, gamma=1, kernel=rbf ......................................
[CV] .......... C=0.1, gamma=1, kernel=rbf, score=0.500, total=   0.5s
[CV] C=0.1, gamma=1, kernel=rbf ......................................
[CV] .......... C=0.1, gamma=1, kernel=rbf, score=0.500, total=   0.5s
[CV] C=0.1, gamma=1, kernel=rbf ......................................
[CV] .......... C=0.1, gamma=1, kernel=rbf, score=0.500, total=   0.5s
[CV] C=0.1, gamma=1, kernel=rbf ......................................
[CV] .......... C=0.1, gamma=1, kernel=rbf, score=0.500, total=   0.5s
[CV] C=0.1, gamma=1, kernel=rbf ......................................
[CV] .

[CV] ............ C=1, gamma=1, kernel=rbf, score=0.700, total=   0.6s
[CV] C=1, gamma=1, kernel=rbf ........................................
[CV] ............ C=1, gamma=1, kernel=rbf, score=0.760, total=   0.6s
[CV] C=1, gamma=1, kernel=rbf ........................................
[CV] ............ C=1, gamma=1, kernel=rbf, score=0.680, total=   0.6s
[CV] C=1, gamma=1, kernel=rbf ........................................
[CV] ............ C=1, gamma=1, kernel=rbf, score=0.730, total=   0.6s
[CV] C=1, gamma=1, kernel=rbf ........................................
[CV] ............ C=1, gamma=1, kernel=rbf, score=0.715, total=   0.6s
[CV] C=1, gamma=10, kernel=rbf .......................................
[CV] ........... C=1, gamma=10, kernel=rbf, score=0.500, total=   0.6s
[CV] C=1, gamma=10, kernel=rbf .......................................
[CV] ........... C=1, gamma=10, kernel=rbf, score=0.500, total=   0.6s
[CV] C=1, gamma=10, kernel=rbf .......................................
[CV] .

[CV] .......... C=10, gamma=10, kernel=rbf, score=0.500, total=   0.7s
[CV] C=10, gamma=10, kernel=rbf ......................................
[CV] .......... C=10, gamma=10, kernel=rbf, score=0.500, total=   0.7s
[CV] C=10, gamma=10, kernel=rbf ......................................
[CV] .......... C=10, gamma=10, kernel=rbf, score=0.505, total=   0.7s
[CV] C=10, gamma=10, kernel=rbf ......................................
[CV] .......... C=10, gamma=10, kernel=rbf, score=0.505, total=   0.6s
[CV] C=10, gamma=10, kernel=rbf ......................................
[CV] .......... C=10, gamma=10, kernel=rbf, score=0.505, total=   0.7s
[CV] C=10, gamma=10, kernel=rbf ......................................
[CV] .......... C=10, gamma=10, kernel=rbf, score=0.505, total=   0.7s
[CV] C=10, gamma=10, kernel=rbf ......................................
[CV] .......... C=10, gamma=10, kernel=rbf, score=0.505, total=   0.7s
[CV] C=10, gamma=100, kernel=rbf .....................................
[CV] .

[CV] ........ C=100, gamma=100, kernel=rbf, score=0.500, total=   1.3s
[CV] C=100, gamma=100, kernel=rbf ....................................
[CV] ........ C=100, gamma=100, kernel=rbf, score=0.500, total=   1.3s
[CV] C=100, gamma=100, kernel=rbf ....................................
[CV] ........ C=100, gamma=100, kernel=rbf, score=0.500, total=   1.4s
[CV] C=100, gamma=100, kernel=rbf ....................................
[CV] ........ C=100, gamma=100, kernel=rbf, score=0.500, total=   1.4s
[CV] C=100, gamma=100, kernel=rbf ....................................
[CV] ........ C=100, gamma=100, kernel=rbf, score=0.505, total=   1.4s
[CV] C=100, gamma=100, kernel=rbf ....................................
[CV] ........ C=100, gamma=100, kernel=rbf, score=0.505, total=   1.4s
[CV] C=100, gamma=100, kernel=rbf ....................................
[CV] ........ C=100, gamma=100, kernel=rbf, score=0.505, total=   1.4s
[CV] C=100, gamma=100, kernel=rbf ....................................
[CV] .

[Parallel(n_jobs=1)]: Done 250 out of 250 | elapsed:  2.7min finished


IS Best Score Sharpe Ratio: 21.373090
Best IS Sharpe ratio: 201.000000
Lowest IS Sharpe Ratio: 8.064150
Mean Sharpe Ratio: 104.225012


In [10]:
in_sample_sharpe_ratio(Rclf)

IS Best Score Sharpe Ratio: 2.509210
Best IS Sharpe ratio: inf
Lowest IS Sharpe Ratio: 2.459825
Mean Sharpe Ratio: inf


C:\Users\Wei_X\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: RuntimeWarning: divide by zero encountered in double_scalars
  """


In [11]:
Rclf1 = RandomizedSearchCV(estimator = clf_svm, 
                          param_distributions = param_distributions,
                          n_iter = 25, # newly added as well
                          scoring='accuracy', 
                          n_jobs=None, 
                          #iid='deprecated', take note of depreciation! version 0.22 warning
                          refit=True,
                          cv= n_splits, # 10 fold CV, but! stratified KFold
                          verbose=10)

Rclf1.fit(X,
         y['bin'],
         sample_weight = None) #sample_weight here!!

in_sample_sharpe_ratio(Rclf1)

Fitting 10 folds for each of 25 candidates, totalling 250 fits
[CV] C=0.15672202039716004, gamma=27.760222656384236 .................


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV]  C=0.15672202039716004, gamma=27.760222656384236, score=0.500, total=   1.2s
[CV] C=0.15672202039716004, gamma=27.760222656384236 .................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    1.1s remaining:    0.0s


[CV]  C=0.15672202039716004, gamma=27.760222656384236, score=0.500, total=   1.2s
[CV] C=0.15672202039716004, gamma=27.760222656384236 .................


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    2.3s remaining:    0.0s


[CV]  C=0.15672202039716004, gamma=27.760222656384236, score=0.500, total=   1.2s
[CV] C=0.15672202039716004, gamma=27.760222656384236 .................


[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:    3.5s remaining:    0.0s


[CV]  C=0.15672202039716004, gamma=27.760222656384236, score=0.500, total=   1.2s
[CV] C=0.15672202039716004, gamma=27.760222656384236 .................


[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed:    4.7s remaining:    0.0s


[CV]  C=0.15672202039716004, gamma=27.760222656384236, score=0.500, total=   1.3s
[CV] C=0.15672202039716004, gamma=27.760222656384236 .................


[Parallel(n_jobs=1)]: Done   5 out of   5 | elapsed:    6.0s remaining:    0.0s


[CV]  C=0.15672202039716004, gamma=27.760222656384236, score=0.505, total=   1.2s
[CV] C=0.15672202039716004, gamma=27.760222656384236 .................


[Parallel(n_jobs=1)]: Done   6 out of   6 | elapsed:    7.2s remaining:    0.0s


[CV]  C=0.15672202039716004, gamma=27.760222656384236, score=0.505, total=   1.2s
[CV] C=0.15672202039716004, gamma=27.760222656384236 .................


[Parallel(n_jobs=1)]: Done   7 out of   7 | elapsed:    8.4s remaining:    0.0s


[CV]  C=0.15672202039716004, gamma=27.760222656384236, score=0.505, total=   1.1s
[CV] C=0.15672202039716004, gamma=27.760222656384236 .................


[Parallel(n_jobs=1)]: Done   8 out of   8 | elapsed:    9.5s remaining:    0.0s


[CV]  C=0.15672202039716004, gamma=27.760222656384236, score=0.505, total=   1.1s
[CV] C=0.15672202039716004, gamma=27.760222656384236 .................


[Parallel(n_jobs=1)]: Done   9 out of   9 | elapsed:   10.7s remaining:    0.0s


[CV]  C=0.15672202039716004, gamma=27.760222656384236, score=0.505, total=   1.1s
[CV] C=57.55419286589449, gamma=0.43792557443306657 ..................
[CV]  C=57.55419286589449, gamma=0.43792557443306657, score=0.925, total=   0.6s
[CV] C=57.55419286589449, gamma=0.43792557443306657 ..................
[CV]  C=57.55419286589449, gamma=0.43792557443306657, score=0.955, total=   0.6s
[CV] C=57.55419286589449, gamma=0.43792557443306657 ..................
[CV]  C=57.55419286589449, gamma=0.43792557443306657, score=0.970, total=   0.6s
[CV] C=57.55419286589449, gamma=0.43792557443306657 ..................
[CV]  C=57.55419286589449, gamma=0.43792557443306657, score=0.955, total=   0.6s
[CV] C=57.55419286589449, gamma=0.43792557443306657 ..................
[CV]  C=57.55419286589449, gamma=0.43792557443306657, score=0.930, total=   0.6s
[CV] C=57.55419286589449, gamma=0.43792557443306657 ..................
[CV]  C=57.55419286589449, gamma=0.43792557443306657, score=0.850, total=   0.6s
[CV] C

[CV]  C=2.484851414405741, gamma=0.019904226105394024, score=0.975, total=   0.3s
[CV] C=2.484851414405741, gamma=0.019904226105394024 .................
[CV]  C=2.484851414405741, gamma=0.019904226105394024, score=0.940, total=   0.3s
[CV] C=2.484851414405741, gamma=0.019904226105394024 .................
[CV]  C=2.484851414405741, gamma=0.019904226105394024, score=0.850, total=   0.2s
[CV] C=2.484851414405741, gamma=0.019904226105394024 .................
[CV]  C=2.484851414405741, gamma=0.019904226105394024, score=0.885, total=   0.2s
[CV] C=2.484851414405741, gamma=0.019904226105394024 .................
[CV]  C=2.484851414405741, gamma=0.019904226105394024, score=0.845, total=   0.2s
[CV] C=2.484851414405741, gamma=0.019904226105394024 .................
[CV]  C=2.484851414405741, gamma=0.019904226105394024, score=0.875, total=   0.3s
[CV] C=2.484851414405741, gamma=0.019904226105394024 .................
[CV]  C=2.484851414405741, gamma=0.019904226105394024, score=0.860, total=   0.2s


[CV]  C=3.998115813506231, gamma=0.6085977830008019, score=0.820, total=   0.6s
[CV] C=3.998115813506231, gamma=0.6085977830008019 ...................
[CV]  C=3.998115813506231, gamma=0.6085977830008019, score=0.840, total=   0.6s
[CV] C=3.998115813506231, gamma=0.6085977830008019 ...................
[CV]  C=3.998115813506231, gamma=0.6085977830008019, score=0.845, total=   0.7s
[CV] C=0.42886643804112307, gamma=0.01008651146927285 ................
[CV]  C=0.42886643804112307, gamma=0.01008651146927285, score=0.920, total=   0.3s
[CV] C=0.42886643804112307, gamma=0.01008651146927285 ................
[CV]  C=0.42886643804112307, gamma=0.01008651146927285, score=0.925, total=   0.3s
[CV] C=0.42886643804112307, gamma=0.01008651146927285 ................
[CV]  C=0.42886643804112307, gamma=0.01008651146927285, score=0.970, total=   0.3s
[CV] C=0.42886643804112307, gamma=0.01008651146927285 ................
[CV]  C=0.42886643804112307, gamma=0.01008651146927285, score=0.975, total=   0.3s
[C

[CV]  C=0.3847152048096837, gamma=1.5624727057871528, score=0.500, total=   0.5s
[CV] C=0.3847152048096837, gamma=1.5624727057871528 ..................
[CV]  C=0.3847152048096837, gamma=1.5624727057871528, score=0.500, total=   0.6s
[CV] C=0.3847152048096837, gamma=1.5624727057871528 ..................
[CV]  C=0.3847152048096837, gamma=1.5624727057871528, score=0.500, total=   0.5s
[CV] C=0.3847152048096837, gamma=1.5624727057871528 ..................
[CV]  C=0.3847152048096837, gamma=1.5624727057871528, score=0.500, total=   0.6s
[CV] C=0.3847152048096837, gamma=1.5624727057871528 ..................
[CV]  C=0.3847152048096837, gamma=1.5624727057871528, score=0.505, total=   0.6s
[CV] C=0.3847152048096837, gamma=1.5624727057871528 ..................
[CV]  C=0.3847152048096837, gamma=1.5624727057871528, score=0.505, total=   0.5s
[CV] C=0.3847152048096837, gamma=1.5624727057871528 ..................
[CV]  C=0.3847152048096837, gamma=1.5624727057871528, score=0.505, total=   0.6s
[CV] C=

[CV]  C=0.04463639175731548, gamma=0.21957197036133624, score=0.825, total=   0.5s
[CV] C=0.04463639175731548, gamma=0.21957197036133624 ................
[CV]  C=0.04463639175731548, gamma=0.21957197036133624, score=0.810, total=   0.5s
[CV] C=0.04463639175731548, gamma=0.21957197036133624 ................
[CV]  C=0.04463639175731548, gamma=0.21957197036133624, score=0.810, total=   0.5s
[CV] C=0.04463639175731548, gamma=0.21957197036133624 ................
[CV]  C=0.04463639175731548, gamma=0.21957197036133624, score=0.830, total=   0.5s
[CV] C=0.04463639175731548, gamma=0.21957197036133624 ................
[CV]  C=0.04463639175731548, gamma=0.21957197036133624, score=0.795, total=   0.5s
[CV] C=0.047510626135955775, gamma=0.07612942514991188 ...............
[CV]  C=0.047510626135955775, gamma=0.07612942514991188, score=0.915, total=   0.4s
[CV] C=0.047510626135955775, gamma=0.07612942514991188 ...............
[CV]  C=0.047510626135955775, gamma=0.07612942514991188, score=0.925, total

[Parallel(n_jobs=1)]: Done 250 out of 250 | elapsed:  2.6min finished


IS Best Score Sharpe Ratio: 19.242468
Best IS Sharpe ratio: 219.745415
Lowest IS Sharpe Ratio: 9.359770
Mean Sharpe Ratio: 106.661255


### Scoring Methods Comparison

To ensure fairness in comparing across both methods and scoring type. We will only take the "best" score given provided and lowest score that we can find.

When using RSCV, remember to provide random_state. Otherwise, your answers will keep changing.

GSCV (Neg_log_loss):

In-Sample Best Score Sharpe Ratio: 2.487544 (Lowest)

Lowest In-Sample Sharpe Ratio: 2.194502

GSCV (Accuracy):

In-Sample Best Score Sharpe Ratio: 21.373090

Lowest In-Sample Sharpe Ratio: 8.064150

RSCV (Neg_log_loss):

In-Sample Best Score Sharpe Ratio: 2.509210

Lowest In-Sample Sharpe Ratio: 2.175036 (Lowest)

RSCV (Accuracy):

In-Sample Best Score Sharpe Ratio: 23.442070

Lowest In-Sample Sharpe Ratio: 6.252751

**At-a-glance**

GridSearchCV (Neg_log_loss) produces a lowest sharpe ratio based on the best score provided.

However, when comparing with RSCV (Neg_log_loss) lowest in-sample sharpe ratio. It seems higher.

This demostrates a consistence in sharpe ratio.

This characteristic is carried over when using "accuracy" as a scoring method (RSCV has a lower IS SR: 6.252751).

When comparing between "accuracy" and "neg_log_loss", it is very obvious accuracy provides a very high sharpe ratio.

But it is even more obvious "accuracy" has a large difference between best and lowest SR.

It seems that "accuracy" as a scoring method is not very accurate.

This is also mentioned in AFML page 134, section 9.4.

### Conclusion

In short, strongly encourage cross-entropy (neg_log_loss) as a scoring method.

Refer to the below negative log loss formula (Binary classes):

![logo](http://wiki.fast.ai/images/2/28/Cross_entropy_formula.png)
    
For natural logarithm if p was to be < 1. The answer would always be negative. Hence to make it let less confusing and comparable to other metrics, we adopt negative sign (Minimize negative log loss).

If you maximise log loss instead (Remove negative sign only/ Maximum log likelihood), your probability would be opposite of the below. Your prediction probability will keep increasing rapidly (Instead of penalizing errorrous prediction, it will end up endorsing such mistakes).

![log_loss](http://wiki.fast.ai/images/4/43/Log_loss_graph.png)

When you have a strategy that bets equally. Accuracy may prove to be bad choice since they attribute true prediction with erroreous prediction equally (your predicted value equals the actual value). 

Under Kelly's criterion, if your lose rate is 50/50 it will lead to a quick ruination.(Recall 2p-1)

If sample weights were included, accuracy may be considered. In most cases cross-entropy is a better choice, since it will factor in the uncertainty based on how much it varies from actual label.

Update: Up till this stage, I realised Dr Marco teaches deep technical knowledge with useful resources to develop further understanding. But he only ask key conceptual questions.